### PR TITLE
CELDEV-965 - Sitemap reordering fails with JS exception

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/CELDEV-965

### CODE REVIEW

File: src/main/webapp/resources/celJS/navigation/reorderMenuItems.js [hunk 1]
```diff
@@ -18,344 +18,340 @@
...
-
-CELEMENTS.reorder.DDReorder.prototype = {
+(function(window, undefined) {
+  "use strict";
+
+  /**
+   * Navigation Reordering
+   */
+  if (typeof window.CELEMENTS == "undefined") { window.CELEMENTS = {}; }
+  if (typeof window.CELEMENTS.reorder == "undefined") { window.CELEMENTS.reorder = {}; }
```
Comment (msladek) [inline src/main/webapp/resources/celJS/navigation/reorderMenuItems.js]:
```suggestion
  if (typeof window.CELEMENTS === "undefined") { window.CELEMENTS = {}; }
  if (typeof window.CELEMENTS.reorder === "undefined") { window.CELEMENTS.reorder = {}; }
```

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#set($menuItemObj = $menuItemDoc.getObject('Celements2.MenuItem'))
+#set($menuItem_parentDoc =  $menuItemDoc.getParent())
+#set($menuPart = $menuItemObj.getProperty('part_name').getValue())
+#else
+#if("$!overwriteWeb" == '')
+#set($overwriteWeb = $doc.getSpace())
+#end
+#set($menuPart = $elemParams.get(1))
+#end
+#set($l = $doc.getURL('view', 'xpage=overlay&conf=AddPage'))
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm]:
```suggestion
#set($l = $services.url.getURL($doc.documentReference, 'view', 'xpage=overlay&conf=AddPage'))
```

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#set($l = "${l}&space=${overwriteWeb}&doc_parent=$!{menuItem_parentDoc}")
+#set($l = "${l}&menuPart=$!{menuPart}&pageTypeCat=celltype")
+#set($cmenu_pageLink = "${l}&menu_position=&language=${default_language}&redirmode=editMenu")
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm]: hand the entire query string to `services.url.getURL(...)`

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#if(("$!elemId" != '') && ($services.treeNode.getMaxConfiguredNavigationLevel() > $celementsweb.getDocumentParentsList($elemId, true).size()))
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm]: use `WebUtilsScriptService.getDocumentParentsDocRefList(DocumentReference, boolean)` instead.
consider extracting variables for better readability.

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#if(("$!elemId" != '') && ($services.treeNode.getMaxConfiguredNavigationLevel() > $celementsweb.getDocumentParentsList($elemId, true).size()))
+#set($overwriteWeb = '')
+#set($menuItemDocRef = $services.model.resolveDocument($elemId))
+#if($elemParams.size() > 1)
+#set($overwriteWeb = $elemParams.get(1))
+#end
+#if("$!overwriteWeb" == '')
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm:9]: always true, remove `if` or line#2

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#set($overwriteWeb = $elemParams.get(1))
+#end
+#if("$!overwriteWeb" == '')
+#set($overwriteWeb = "$!menuItemDocRef.getLastSpaceReference().getName()")
+#end
+#set($overwriteWeb = "${overwriteWeb}-$!menuItemDocRef.getName()")
+#set($l = $doc.getURL('view', 'xpage=overlay&conf=AddPage'))
+#set($l = "${l}&space=${overwriteWeb}&doc_parent=${elemId}")
+#set($l = "${l}&language=${default_language}&redirmode=newpageedit")
+#set($cmenu_pageLink = "${l}&xredirect=$escapetool.url($doc.getURL('view'))")
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm]: same points as above

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/del_cell.vm [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#if("$!elemId" != '')
+#set($menuItemDoc = $services.modelAccess.getDocument($services.model.resolveDocument($elemId)))
+#if(${elemId} == ${doc.fullName})
+  #set($red = "${doc.web}.WebHome")
+#else
+  #set($red = ${elemId})
+#end
+#set($l = "confirmURL(&quot;$adminMsg.get('cel_cell_node_confirm_delete')&quot;,")
+#set($params = "xpage=celements_ajax&ajax_mode=delMenuItem&space=${doc.space}&red=${red}")
+#set($l = "${l}&quot;$xwiki.getURL($elemId,'view', $params)")
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/del_cell.vm]: use `services.url`

File: src/main/webapp/templates/celMacros/SitemapForSpace.vm [hunk 8]
```diff
@@ -52,9 +52,10 @@
 <div id="spaceNavigation">
 	<label for="spaceSelect">Bereich wechseln zu:</label>
 	<select id="spaceSelector" name="spaceSelect">
-#foreach($spaceName in $spaceList)
-	#getDefaultParams($spaceName)
-		<option label="$spaceName" value="$doc.getURL('view', $defaultParams)">$spaceName</option>
+#foreach($spaceNameSelect in $spaceList)
+	#getDefaultParams($spaceNameSelect)
+		<option label="$spaceNameSelect" value="$doc.getURL('view', $defaultParams)"##
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/SitemapForSpace.vm]:
```suggestion
  #getDefaultParams($spaceNameSelect)
    <option label="$spaceNameSelect" value="$services.url.getURL($doc.documentReference, 'view', $defaultParams)"##
```

File: src/main/webapp/templates/celMacros/SitemapForSpace.vm [hunk 9]
```diff
@@ -52,9 +52,10 @@
...
+#if("$!spaceName" == "$!spaceNameSelect") selected #end>$spaceName</option>
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/SitemapForSpace.vm]:
```suggestion
#if("$!spaceName" == "$!spaceNameSelect") selected #end>$spaceNameSelect</option>

```

**Review by msladek — CHANGES_REQUESTED** [review]

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm (hunk 5 above)
Comment (fpichler) [reply src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm:9]: don't see why this is always true. If line 5 (now lin 7) sets $overwriteWeb to elemParams 1 which could be non empty.

Comment (msladek) [reply src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm:9]: yes, my bad

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#if("$!overwriteWeb" == '')
+#set($overwriteWeb = $doc.getSpace())
+#end
+#set($menuPart = $elemParams.get(1))
+#end
+#set($l = 'xpage=overlay&conf=AddPage')
+#set($l = "${l}&space=${overwriteWeb}&doc_parent=$!{menuItem_parentDoc}")
+#set($l = "${l}&menuPart=$!{menuPart}&pageTypeCat=celltype")
+#set($l = "${l}&menu_position=&language=${default_language}&redirmode=editMenu")
+#set($cmenu_pageLink = $services.url.getURL($doc.documentReference, 'view', "{l}"))
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_cell.vm]:
```suggestion
#set($cmenu_pageLink = $services.url.getURL($doc.documentReference, 'view', "${l}"))
```

File: src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+#set($overwriteWeb = $elemParams.get(1))
+#end
+#if("$!overwriteWeb" == '')
+#set($overwriteWeb = "$!menuItemDocRef.getLastSpaceReference().getName()")
+#end
+#set($overwriteWeb = "${overwriteWeb}-$!menuItemDocRef.getName()")
+#set($l = 'xpage=overlay&conf=AddPage')
+#set($l = "${l}&space=${overwriteWeb}&doc_parent=${elemId}")
+#set($l = "${l}&language=${default_language}&redirmode=newpageedit")
+#set($l = "${l}&xredirect=$escapetool.url($doc.getURL('view'))")
```
Comment (msladek) [inline src/main/webapp/templates/celContextMenu/cel_cm_presentationeditor_nodes/add_sub_cell.vm]:
```suggestion
#set($l = "${l}&xredirect=$escapetool.url($services.url.getURL($doc.documentReference, 'view'))")
```

**Review by msladek — CHANGES_REQUESTED** [review]

**Review by msladek — APPROVED** [review]

**Pull request merged**